org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/junit/launcher/AdvancedJUnitLaunchConfigurationDelegateTest.java

```diff
@@ -0,0 +1,200 @@
+		String projectName= "JUnitLaunchConfigurationDelegate-TestProject-MultiplePackagesInOneFolder";
+		fJavaProject= JavaProjectHelper.createJavaProject(projectName, "bin");
+		String firstTestSrcFolder= "test-src1";
+		IPackageFragmentRoot testSrc1= JavaProjectHelper.addSourceContainer(fJavaProject, firstTestSrcFolder);
+		String testPackage= "p1";
+		IPackageFragment packageTestSrc1= testSrc1.createPackageFragment(testPackage, true, null);
+		String contentsTestSrc1Test= """
+				public class FirstTest {
+				@org.junit.jupiter.api.Test
+				public void myTest() { }
+				}
+				""";
+		String secondTestPackage= "p2";
+		packageTestSrc1.createCompilationUnit("FirstTest.java", contentsTestSrc1Test, true, null);
+		IPackageFragment packageTestSrc2= testSrc1.createPackageFragment(secondTestPackage, true, null);
+		String contentsTestSrc2Test= """
+				public class SecondTest {
+				@org.junit.jupiter.api.Test
+				public void mySecondTest() { }
+				}
+				""";
+		packageTestSrc2.createCompilationUnit("SecondTest.java", contentsTestSrc2Test, true, null);
+
+		List<String> fileLines= showCommandLineAndExtractContentOfTestNameFile(projectName, fJavaProject, testSrc1);
+		String lineForFirstTest= "p1.FirstTest";
+		String lineForSecondTest= "p2.SecondTest";
+		assertThat(fileLines).contains(lineForFirstTest, lineForSecondTest).size().isEqualTo(2);
+	}
+
+	@Test
+	public void runTestsInSourceFolderHandlesMultipleSourceFilesInPackagesInASourceFolderCorrectly() throws Exception {
+		String projectName= "JUnitLaunchConfigurationDelegate-TestProject-MultipleSourceFilesInOnePackage";
+		fJavaProject= JavaProjectHelper.createJavaProject(projectName, "bin");
+		String firstTestSrcFolder= "test-src1";
+		IPackageFragmentRoot testSrc1= JavaProjectHelper.addSourceContainer(fJavaProject, firstTestSrcFolder);
+		String testPackage= "p1";
+		IPackageFragment packageTestSrc1= testSrc1.createPackageFragment(testPackage, true, null);
+		String contentsTestSrc1Test= """
+				public class FirstTest {
+				@org.junit.jupiter.api.Test
+				public void myTest() { }
+				}
+				""";
+		packageTestSrc1.createCompilationUnit("FirstTest.java", contentsTestSrc1Test, true, null);
+		String contentsTestSrc2Test= """
+				public class SecondTest {
+				@org.junit.jupiter.api.Test
+				public void mySecondTest() { }
+				}
+				""";
+		packageTestSrc1.createCompilationUnit("SecondTest.java", contentsTestSrc2Test, true, null);
+
+		List<String> fileLines= showCommandLineAndExtractContentOfTestNameFile(projectName, fJavaProject, testSrc1);
+		String lineForFirstTest= "p1.FirstTest";
+		String lineForSecondTest= "p1.SecondTest";
+		assertThat(fileLines).contains(lineForFirstTest, lineForSecondTest).size().isEqualTo(2);
+	}
+
+	@Test
+	public void runTestsInSourceFolderOnlyUsesTopLevelClasses() throws Exception {
+		String projectName= "JUnitLaunchConfigurationDelegate-TestProject-OnlyTopLevelClasses";
```

Comment:
@erik-brangs 

It looks like there was an explicit decision to exclude nested test classes. Do you remember why nested (and inner) classes are being excluded with this change? I think this is a regression because this change means that a bunch of the CDT tests now don't run. I put some examples in https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2591#issuecomment-3482798331 to compare JUnit4 and 5 behaviour and across IDE versions as needed.

If you have any thoughts on this I would appreciate your insights on #2591